lib/index.coffee

```diff
@@ -76,31 +105,38 @@ class DockerGC
 				@currentMtimes = layer_mtimes
 
 	removeImage: (image) ->
-		return this.tryRemoveImageBy(image, image.repoTags) ||
-					 this.tryRemoveImageBy(image, image.repoDigests) ||
-					 this.tryRemoveImageBy(image, [image.id])
+		return this.tryRemoveImageBy(image, image.repoTags, 'tag') ||
+					 this.tryRemoveImageBy(image, image.repoDigests, 'digest') ||
+					 this.tryRemoveImageBy(image, [image.id], 'id')
 
-	tryRemoveImageBy: (image, attributes) ->
+	tryRemoveImageBy: (image, attributes, removalType) ->
 		if attributes? and attributes.length > 0
 			Promise.each attributes, (attribute) =>
 				console.log("GC: Removing image : #{attribute} (id: #{image.id})")
 				@docker.getImage(attribute).remove(noprune: true)
+				.then ->
+					@metrics?.histogram('gc_images_removed', removalType)
 
 	garbageCollect: (reclaimSpace, attemptAll = false) ->
 		err = null
+		startTime = Date.now()
+		@metrics?.gauge('gc_space_reclaimed', reclaimSpace)
 		dockerImageTree(@docker, @currentMtimes)
 		.then (tree) ->
 			getImagesToRemove(tree, reclaimSpace)
 		.each (image) =>
 			@removeImage(image)
 			.catch (e) ->
+				@metrics?.histogram('gc_image_removal_errors', e.statusCode)
 				console.log('GC: Failed to remove image: ', image)
 				console.log(e)
 				if attemptAll
 					err ?= e
 				else
+					recordGcRunTime(@metrics, startTime)
```

Comment:
Will this `.catch` (L129) block run for *each* of the images in `.each`? Or will it merely catch only the first error and throw?

In the former case this would mean recording several times, confusing the semantics between `recordGcRunTime` being used to time the total job vs timing "time until first, second, third, etc... image errors".

There should be a way to use some kind of "finally" clause to only record this in one line of code which is guaranteed to run once (even if, as written, it's guaranteed, it's not as clear and could get messed up by future changes not aware of the subtleties).

Reply:
It depends on how is called from the Builder. By default is called with the `attemptAll` set to `false`, so it will throw on the first error.  In summary:

* If `attemptAll` == `false` (the default case), the total job time will be the time from moment the GC started till the first error was thrown

* If `attemptAll` == `true`, the catch will save the error but not record the time (L134). Once the `each` is finished, the execution will jump to the `catch` block (L139) where will first record the time and then throw in case there was an error. So in this case, the total job time will be from the moment the GC started till it finished.

Does this make sense? I write this differently, but this GC is a legacy throwaway code: we need to start collecting metrics while I finish the new GC (which is written in Go, I'll bother you for that one later xD)

total job time will be the time from moment the GC started till the first error was thrown

Reply:
Thanks for explaining, I should've just been more careful about reading and reasoning with the code. Yes, you know, this makes sense and is clear from reading it. This can stay how it is